### PR TITLE
Improve performance of `st.domains()`

### LINKED ISSUE #4201: The domain strategy is very slow
I recognise that `domains` is provisional, however, it's in use by the `emails` strategy. I'd prefer to test over a set of valid RFC 3986 domains rather than defining a strategy to naively create some domains. `domains` takes on average 2.6s, which is significantly negatively impacting both the `emails` and `urls` strategies.

This small set of timings shows the differences, I've added instantiating a class as a baseline and the `text` strategy to demonstrate the additional overhead that `domains` is causing

```python
import timeit
from dataclasses import dataclass
from hypothesis import given, strategies as st
import hypothesis.provisional as p

@dataclass
class A:
    value: str

def baseline():
    A(value='test@example.com')
    
@given(st.text())
def text(value: str):
    A(value=value)
    
@given(p.domains())
def domains(value: str):
    A(value=value)

@given(p.urls())
def urls(value: str):
    A(value=value)

@given(st.emails())
def emails(value: str):
    A(value=value)
    
@st.composite
def fake_domain_strategy(draw: st.DrawFn) -> str:
    domain = draw(st.characters(min_codepoint=97, max_codepoint=122))
    return domain + '.com'

@given(fake_domain_strategy())
def fake_domain(value: str):
    A(value=value)

if __name__ == "__main__":

    b = timeit.timeit(baseline, number=1000)
    t = timeit.timeit(text, number=10)
    d = timeit.timeit(domains, number=10)
    u = timeit.timeit(urls, number=10)
    e = timeit.timeit(emails, number=10)
    f = timeit.timeit(fake_domain, number=10)

    print('Field    ', 'Time:   ', 'Compared to:', 'Time difference:')
    print('Baseline ', f'{b:.6f}', 'N/A')
    print('Text     ', f'{t:.6f}', 'baseline    ', f'{t - b:.6f}')
    print('Domains  ', f'{d:.6f}', 'baseline    ', f'{d - b:.6f}')
    print('URLs     ', f'{u:.6f}', 'domains     ', f'{u - d:.6f}')
    print('Emails   ', f'{e:.6f}', 'domains     ', f'{e - d:.6f}')
    print('Fake     ', f'{f:.6f}', 'domains     ', f'{f - d:.6f}')

```
results:
```
Field     Time:    Compared to: Time difference:
Baseline  0.000145 N/A
Text      0.303678 baseline     0.303533
Domains   2.720224 baseline     2.720079
URLs      3.560947 domains      0.840723
Emails    2.867589 domains      0.147365
Fake      0.060368 domains      -2.659856
```
The highlight is that the `domains` strategy regularly takes 2.6-2.7s

In the Time difference column I've attempted to strip out the "cost" of using `domains` to provide an estimate of the timings of `urls` and `emails` in comparison to `text`.

* `text` is 0.3s
* `urls` might be 0.8s if it weren't for domains
* `emails` might be 0.1s if it weren't for domains
* (badly) faking a domain takes the expected length of time, somwhere between not using Hypothesis and the `text` strategy

Looking at this through the profiler it doesn't look like anything in `provisional.py` is the issue, instead it's how the core engine is using the `domains` strategy. Being unfamiliar with the internals of Hypothesis this is quite hard to decipher and I'm hoping that someone in the project might have a quicker idea!

Maybe I'm doing something silly?

Thank you

### PR DESCRIPTION
**Pull request by tybug**
Closes #4201. My benchmark results:

#### Master
```
Field     Time:    Compared to: Time difference:
Baseline  0.000182 N/A
Text      0.353401 baseline     0.353219
Domains   3.109082 baseline     3.108901
URLs      3.952818 domains      0.843736
Emails    3.208230 domains      0.099147
Fake      0.065587 domains      -3.043495
```

#### This PR
```
Field     Time:    Compared to: Time difference:
Baseline  0.000170 N/A
Text      0.363252 baseline     0.363082
Domains   1.905849 baseline     1.905679
URLs      2.744583 domains      0.838734
Emails    2.032297 domains      0.126449
Fake      0.067094 domains      -1.838755
```

### CODE REVIEW

File: hypothesis-python/src/hypothesis/provisional.py [hunk 1]
```diff
@@ -89,34 +89,45 @@ def __init__(
...
+        def recase_randomly(draw, tld):
+            tld = list(tld)
+            changes = draw(st.tuples(*(st.booleans() for _ in range(len(tld)))))
+            for i, change_case in enumerate(changes):
+                if change_case:
+                    tld[i] = tld[i].lower() if tld[i].isupper() else tld[i].upper()
+            return "".join(tld)
+
+        self.domain_strategy = (
             st.sampled_from(get_top_level_domains())
```
Comment (tybug) [inline hypothesis-python/src/hypothesis/provisional.py:118]: The previous flatmap dynamically created a lot of uncacheable strategies. variable-length tuples aren't enormously cacheable in general, but they are decently so when the length is the only thing that can vary.
```
            .flatmap(
                lambda tld: st.tuples(
                    *(st.sampled_from([c.lower(), c.upper()]) for c in tld)
                ).map("".join)
            )
```

File: hypothesis-python/src/hypothesis/provisional.py [hunk 2]
```diff
@@ -89,34 +99,37 @@ def __init__(
...
+
         # 1 - Select a valid top-level domain (TLD) name
         # 2 - Check that the number of characters in our selected TLD won't
         # prevent us from generating at least a 1 character subdomain.
         # 3 - Randomize the TLD between upper and lower case characters.
-        domain = data.draw(
+
+        self.domain_strategy = (
             st.sampled_from(get_top_level_domains())
             .filter(lambda tld: len(tld) + 2 <= self.max_length)
```
Comment (Stranger6667) [inline hypothesis-python/src/hypothesis/provisional.py:121]: wondering if it makes sense to make a list of tlds with appropriate lengths before `st.sampled_from` and then avoid the `.filter` call altogether. I guess if `self.max_length` does not change, then it should work, though not sure about performance implications.

Comment (tybug) [reply hypothesis-python/src/hypothesis/provisional.py:121]: After consideration I don't think holding that list in memory is worth avoiding a single filter call here, and also probably degrades quality due to losing retries in `FilteredStrategy`. (`self.max_length` is fixed at strategy definition, so no worries there.)

**Review by Zac-HD — APPROVED** [review]
Thanks, @tybug!

**Pull request merged**